### PR TITLE
Code review n.2 code by Marco Cognolato for Barca and Baù

### PR DESCRIPTION
**Pull request by Cogno-Marco**
## Premise
As per title, this is the code for the second review of Marco Cognolato's code, this time for Barca and Baù.

Review is of 4 files:
- ConnectionHandler
- TableUpdateHandler
- TableUpdateIterator
- KademliaId

In this last one the review is **NOT** of the whole class, but just of the 3 methods that you can see.

### 1. ConnectionHandler
This class has only the role of establishing a connection between 2 users (basically the bootstrap problem)

### 2. KademliaId
Here I've improved/rewritten the code of 3 functions:
- The constructor using a String,
Where the KademliaId is constructed from a string,
- generateNodeIdByDistance,
where a new Id is created from the one of the object
- getFirstSetBitIndex,
where a special distance is calculated between 2 ids and used in other places (like the RoutingTable)

These functions are expecially low level due to the nature of the checks they have to do to the id, and this is one of the reasons why there's probably less code to review than last time.

### 3. TableUpdateHandler and TableUpdateIterator

Here I've used an Iterator Design Pattern to implement an algorithm to update the routing table.
You'll probably notice that the TableUpdateHandler class is static, and there's a reason for that, but I want you to think about why that is before telling you 😉 

___
Happy Holidays and good review! 🎉

### CODE REVIEW

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 1]
```diff
@@ -52,7 +52,33 @@ public KademliaId(SMSPeer peer) {
         }
     }
 
+    /**
+     * Construct the NodeId from a string
+     *
+     * @param data The user generated key string
+     * @throws IllegalArgumentException if data has not the right fit
```
Comment (giobbarca) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: You should better clarify when the exception is thrown (class users should understand from JavaDocs withoud reading the code)

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 2]
```diff
@@ -52,7 +52,33 @@ public KademliaId(SMSPeer peer) {
...
+     * @author Marco Cognolato
+     */
+    public KademliaId(String data) {
+        this.keyBytes = data.getBytes();
```
Comment (giobbarca) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: I would put the code in this method in another method (excluding the `for `cycle) in such a way you don't have replicated algorithm in both constructors

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 3]
```diff
@@ -52,7 +52,33 @@ public KademliaId(SMSPeer peer) {
...
+                keyBytes[i] = (byte) j;
+            }
+        }
+        if (this.keyBytes.length > ID_LENGTH_BYTES) {
+            throw new IllegalArgumentException("Specified Data need to be " + ID_LENGTH_BYTES + " characters long.");
+        }
 
+        //if the string it's too short, add some leading null bytes
+        if(this.keyBytes.length < ID_LENGTH_BYTES){
+            this.keyBytes = addTrailingZeros(this.keyBytes);
```
Comment (giobbarca) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: Shouldn't this method be called `addLeadingZeros(...)`?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 4]
```diff
@@ -146,7 +172,48 @@ public KademliaId xor(KademliaId nid) {
...
+    }
 
+    /**
+     * Counts the number of leading 0's in this NodeId
+     *
+     * @return Integer representing the number of leading 0's,
+     * also returns -1 if there's only leading 0's
+     * @author EdoardoRaimondi, Marco Cognolato
+     */
+    public int getFirstSetBitIndex() {
```
Comment (giobbarca) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:218]: I would called this method something like `getLeadingZerosCount`.
As it is called now, this method name suggest returns the index of the first set bit index.

File: webdictionary/src/main/java/com/eis0/kademlianetwork/TableUpdateIterator.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ * @author Marco Cognolato
+ */
+public class TableUpdateIterator {
+    private int currentCount;
+    private int maxCount;
+    private KademliaId netId;
+    private SMSKademliaRoutingTable table;
+    private SMSPeer netPeer;
+
+    public TableUpdateIterator(int maxCount, KademliaId netId, SMSKademliaRoutingTable table, SMSPeer netPeer) {
```
Comment (giobbarca) [inline webdictionary/src/main/java/com/eis0/kademlianetwork/TableUpdateIterator.java]: Add some checks about `maxCount`, `netId` (if not already done in the `KademliaId` class), `table` and `netPeer`:

- What happens if `maxCount `is 0 or negative?
- Is `netId `always a valid ID?
- What happens if `table` is null? You could catch this "event" in advance and avoid useless code execution.
- Same thing for `netPeer`.

**Review by giobbarca — COMMENTED** [review]
I think it's a good idea to use handler and iterator design patterns in this implementation of Kademlia.
However you specified in the code comments that the iterator is specific to one single routingTable in such a way you can run multiple iterator "instances" at the same time.
I would take advantage of this feature to allow multiple networks management (I think it isn't possible at this time because you're using a singleton design pattern for the `KademliaNetwork` class - not given in these files but "logically obtainable" by the `KademliaNetwork.getInstance()` in rows 75-76 of `ConnectionHandler.java`).

Great coding overall. By the way I suggest some improvements about things I reviewed in detail and a more detailed explanation of methods tasks in the JavaDocs.

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 4 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:218]: > As it is called now, this method name suggest returns the index of the first set bit index.

That's exactly what it does, it returns the index of the first bit = 1. I guess I'll update the specifications to avoid confusion 👍

File: webdictionary/src/main/java/com/eis0/kademlianetwork/TableUpdateIterator.java (hunk 5 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/TableUpdateIterator.java]: Awesome suggestion!
Working on it 👍

**Comment by Cogno-Marco** [comment]
> However you specified in the code comments that the iterator is specific to one single routingTable in such a way you can run multiple iterator "instances" at the same time.
I would take advantage of this feature to allow multiple networks management

Great job for spotting that!
As I said in the Premise 
> You'll probably notice that the TableUpdateHandler class is static, and there's a reason for that

And that's exactly why!
We decided to use a singleton since it's easy to build upon, but we worked with the intention of removing that singleton and having multiple network instances at the same time.
Due to that "constraint" I decided to have TableUpdateHandler a static class (so it can work with any instance of a KademliaNetwork without keeping track of any network-related data).
Basically the Iterator is specific to a single network (since it has to update the table of that specific network) while TableUpdateHandler only has to manage multiple iterators without knowing anything of any network instance. This makes it reusable for any network.

Basically all this work is preparation for upcoming changes which we are not yet ready to do, but at least we have a foundation to build upon.

**Review by giobbarca — COMMENTED** [review]
Now is better! Good job.

> Here I've used an Iterator Design Pattern to implement an algorithm to update the routing table.

I really appreciate how you all plan design patterns before writing code, honestly still difficult for me. It's nice to learn this way too. Thanks!

**Comment by francescobau** [comment]
I really like your way to do the description!!
I hope you wouldn't mind if I take inspiration by it...

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 6]
```diff
@@ -52,7 +52,31 @@ public KademliaId(SMSPeer peer) {
...
+     */
+    public KademliaId(String data) {
+        // if it's in hex form, convert to byte array
+        if(data.matches("[0-9A-F]{" + ID_LENGTH_BYTES*2+"}")){
+            this.keyBytes = new byte[ID_LENGTH_BYTES];
+            for (int i = 0; i < keyBytes.length; i++) {
+                int index = i * 2;
+                int j = Integer.parseInt(data.substring(index, index + 2), 16);
+                keyBytes[i] = (byte) j;
+            }
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:80]: Here I am seeing many magic numbers and magic strings... Is this a hardwiring issue?
Pardon me for my ignorance, but I can't understand the meaning of these things:
- "[0-9A-F]" -> does that mean the identification of a hex number?
- 16 must be meaning "16 bits' length", right?
- I can see that 2 everywhere... But I don't understand what's its purpose in each row. Do you mind to tell me their purpose, please?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 7]
```diff
@@ -80,7 +115,7 @@ public KademliaId(byte[] bytes) {
      * @return A byte array with the number trailed with zeros
      * @author Marco Cognolato
      */
-    private byte[] addTrailingZeros(byte[] number){
+    private byte[] addLeadingZeros(byte[] number){
         //by default each element is set to 0x00
         byte[] toReturn = new byte[ID_LENGTH_BYTES];
         //add to the end (ax expected) each element
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:125]: Again, I'm sorry for my ignorance, and I hope not to bother you by asking these things...
- Row 139: Possible hardwiring? What's that 1 about?
- Rows 161-164: what's the meaning of 7 and 83?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 8]
```diff
@@ -146,7 +181,48 @@ public KademliaId xor(KademliaId nid) {
...
+     * @author Edordo Raimondi, improved by Marco Cognolato
+     */
+    public KademliaId generateNodeIdByDistance(int distance) {
+        if(distance < 0 || distance >= ID_LENGTH) throw new IllegalArgumentException();
+
+        byte[] result = keyBytes.clone();
+
+        // calculate the index of the byte and bit to update
+        int byteToUpdateIndex = (distance) / 8;
+        int bitToUpdateIndex = 7 - (distance % 8);
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: 7 and 8... I don't know if we could consider this hardwiring or not...
Is it?
I mean, 8 means "8 bits", right? And 7 is the maximum amount of bits to update, which is 8-1, right?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 9]
```diff
@@ -146,7 +181,48 @@ public KademliaId xor(KademliaId nid) {
...
     }
 
+    /**
+     * Generates a NodeId that is some distance away from this NodeId
+     *
+     * @param distance The index of the first bit that should be different,
+     *                 From 0 to 159, where 0 changes the most significant bit
+     * @return The newly generated NodeId with a given distance from this.
+     * @throws IllegalArgumentException if the distance is < 0 or >= 160
+     * @author Edordo Raimondi, improved by Marco Cognolato
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: Who's "Edordo Raimondi"?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 10]
```diff
@@ -146,7 +181,48 @@ public KademliaId xor(KademliaId nid) {
...
+        result[byteToUpdateIndex] = (byte)(result[byteToUpdateIndex] ^ 0x01<<bitToUpdateIndex);
+        return new KademliaId(result);
+    }
 
+    /**
+     * Returns the index of the first bit set to 1, starting from the
+     * most significant digit
+     *
+     * @return Index of the first bit set to 1, returns -1 if there's no bit set to 1
+     * @author EdoardoRaimondi, Marco Cognolato
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: Who is "EdoardoRaimondi"?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 11]
```diff
@@ -146,7 +181,48 @@ public KademliaId xor(KademliaId nid) {
...
+        for(int byteIndex = 0; byteIndex < keyBytes.length; byteIndex++) {
+            byte currentByte = keyBytes[byteIndex];
+            //8 bits in a byte, from 0 to 7
+            for(int bitIndex = 7; bitIndex >= 0; bitIndex--) {
+                if(((0x01 << bitIndex) & currentByte) == (0x01 << bitIndex)){
+                    return (7-bitIndex) + byteIndex * 8;
+                }
+            }
+        }
+        return -1;
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: Here we can see better the meaning of 7 and 8... But, since you use them plenty of times, how about setting them as constants?
The same thing is for 0x01... I would have called it BYTE_SKIP, or something like that...
And that -1 I see at the bottom, I would have called it something like INVALID_VALUE...

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java [hunk 12]
```diff
@@ -146,7 +181,48 @@ public KademliaId xor(KademliaId nid) {
...
+    }
 
     /**
      * Gets the bit distance from this NodeId to another NodeId
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:232]: Row 238 and 258: Are these possible hardwiring issues?
Row 259: I'm really confused... How does that work?
Rows 269-275: 1 , 2 , 4 , 0x0F , 0XFF ... Aren't these hardwired?
I'm sorry to insist on these things... I'd like to know better about hardwiring, since I'm a noob, and you have more experience than me...

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+     * @param peer The peer to invite to my network
+     */
+    public void inviteToJoin(SMSPeer peer) {
+        String messageRequest = RequestTypes.JoinPermission.ordinal() + "";
+        SMSMessage message = new SMSMessage(peer, messageRequest);
+        SMSManager.getInstance().sendMessage(message);
+    }
+
+    /**
+     * Asks a given valid peer if I can join HIS Kademlia Network
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java]: I can see why you hadn't used something like "Asks a given valid peer to join his Kademlia Network"... It would be confusing, because it could mean "X asks a valid peer Y to join Y's Kademlia Network", or also "X asks a valid peer Y to join X's Kademlia Network", right?
But still, I don't know if it's good practise to use 1st person in the comments...
How about "Source peer asks Destination peer to join Destination peer's Kademlia Network", or simply "Asks a given valid target peer to join target peer's Kademlia Network", or maybe "Requests to join a valid target peer's Kademlia Network"?
It's just a thing that popped out from my mind...

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    public static void sendAcceptRequest(SMSPeer peer) {
+        /*
+         * If both the other peer sent me a request to join his network, or
+         * a request to invite me to his network, I have to update my routing table
+         * from scratch adding him to the table as a contact, then I have to send
+         * an acceptJoin event, so he can do the same thing.
+         * This serves to more strongly fuse both networks by
+         * creating contacts from the other network
+         * */
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java]: I noticed that you tend to use the 1st person... It indeed makes it easier to understand, but I don't know if that's good practise... Is it?

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * Accepts the request sent by an SMSPeer
+     *
+     * @param peer The peer who sent the request
+     */
+    public static void acceptRequest(SMSPeer peer) {
+        /*
+         * If both the other peer sent me a request to join his network, or
+         * a request to invite me to his network, I have to update my routing table
+         * from scratch adding him to the table as a contact
```
Comment (francescobau) [inline webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: Isn't this part of the comment redundant? I mean... It looks like it's copy-pasted from the one above...
I think you could optimize it by putting acceptRequest method before sendAcceptRequest, so you'd be able to avoid lines of redundant comments...
(By the way, I agree that +1 comment is better than -1 comment, though...)

**Review by francescobau — COMMENTED** [review]
Nice code overall, just some @ author issues, a few possible hardwiring issues and maybe a redundant comment...
But the code is well structured, and I can see that you tried hard to make the code understandable even to noobs like me, so well done! 👍

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 9 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: Oh shiet!

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java (hunk 14 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java]: Mmm... I never thought about it now that I think.
I guess it's just personal preference, so I wouldn't know.
¯\_(ツ)_/¯

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java (hunk 15 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: Yes, the comment is copy pasted, though there's a slightly difference which maybe I should point out more clearly.
there's this piece of comment in `sendAcceptRequest` which is not in `acceptRequest`:
> then I have to send an acceptJoin event, so he can do the same thing.

The reason most of the comment is the same is to have every functions self documented, so when I look at how a function works I can stay inside that function, instead of having to look around for the documentation.

I'll change a part of the comment to make that difference pop up more easily, but I think I'll still keep these functions in this order since `sendAcceptRequest` uses `acceptRequest`, so having them in order is (at least in my opinion) better to read

Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: Alright, updated. Is it better now?

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 12 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:232]: I'll explain rows 258 and 259, even though I didn't work on those.
First we create a BigInteger from the byte array (the 1 is a signum value used by the constructor to say how the number should behave. It can be -1 meaning the number is negative, 1 meaning the number is positive, and 0 meaning the whole number is zero)
Then we srite that number as a string, formatting it in the process.
A string formatted for hex numbers is in the form "%0nX", where n is the minimum number of characters in hex form.
so for example in the format "%05X" you can write "0xFFFFF" as well as "0x000AB" since there are 5 characters after the "x"
What we do is, since 1 byte is 2 hex characters, we take the length of the array and "multiply by two", which you can also do if you bitshift 1 time to the left.
Again, I didn't work on that so I don't know why that's done that way.

Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:232]: Regarding rows 269-275 I think I can just remove all that method and simply call `hexRepresentation` since I believe they're doing the same thing in different ways.

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 7 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:125]: As I've explained in another comment the "1" is a value the BigInteger construct wants to say that the number is positive.
Regarding rows 161-164 I don't know why those values have been chosen specifically. But seeing what it does I think I could simply call and return that `Arrays.hashCode(this.keyBytes);`

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 6 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:80]: "[0-9A-F]" is the first part of a regex pattern. It's basically checking if every character is a character between 0-9 and A-F (or in another way to write it, if it's one of these characters "0123456789ABCDEF"), so I'm basically checking if evey character is a Hex representation of a number.
The other thing that pattern is checking (using that `"{" + ID_LENGTH_BYTES*2 +"}"` ) is check if the string is long ID_LENGTH_BYTES * 2, since 2 hex numbers are 1 byte, I need ID_LENGTH_BYTES * 2 to create a byte array ID_LENGTH_BYTES long.

The "16" is not quite "16 bits", it's meaning in the parseInt method is "base 16"

The "2" is used as follows:

the ID_LENGTH_BYTES * 2 has already been adressed ^

inside that for it's used in 2 ways.
2 hex caracters need to be transformed into a single number and then put into the keyBytes array, so:
I iterate over the keyBytes array length,
I take 2 characters at a time from the `data` String,
parse them from base 16 to int
and finally put them into the array at index `i`

So those "2" are:
`int index = i * 2` to convert from index of keyBytes to index of `data` where to take the string
and
`data.substring(index, index + 2)` where I take 2 characters of the `data` string at index `index`

Hope this was more clear

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 8 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: you're right, 8 is "8 bits"
The 7 is 8-1 but because when you count from 0 to 7 there are 8 values.
It's the same thing as "an array of length `n` has indeces from 0 to `n-1`"

I can definitely see why it can be considered hardwiring, though, is it right to put those in a const value?
I mean, a const value might be changed in the future by hand, but would it even make sense?
changing from 8 to for example 10 would mean that a byte is 10 bits, which clearly makes no sense.

That said, it might be better to put it in a const and clearly having a comment saying to not edit it, so I'll do it.

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 11 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java]: As per another comment I will move 7s and 8s into consts, and I'll do the same for the `-1`, but that 0x01 is specific for the operation I'm doing here, which is selecting a specific bit and seeing if it's equal to 1, so I don't think it's such a big hardwiring.

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 6 above)
Comment (francescobau) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:80]: Yes, very useful explanation. Thank you! 👍

File: webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java (hunk 12 above)
Comment (francescobau) [reply webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java:232]: Very nice to know! Thank you for the explanation.

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java (hunk 15 above)
Comment (francescobau) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: Not really... It looks kinda confusing to me... :(
It looks like:
[sendAcceptRequest] When I receive a request to join his network, or (when I receive) a request to join his network I have to send an acceptJoin event, to notify him and let him do his work.
[acceptRequest]  If both the other peer sent me a request to join his network, or (the other peer sent me) a request to invite me to his network, I have to update my routing table from scratch adding him to the table as a contact

Did you mean to say these things above ⬆️ , or these things below ⬇️ ?

[sendAcceptRequest] When I receive a request to join his network, or when I request to join his network I have to send an acceptJoin event, to notify him and let him do his work.
[acceptRequest]  If both the other peer sent me a request to join his network, or I request the other peer to join his network, I have to update my routing table from scratch adding him to the table as a contact

**Review by francescobau — COMMENTED** [review]
Thank you very much for the explanations, but you might consider looking for the comments again...

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java (hunk 14 above)
Comment (francescobau) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java]: Oh I see...

File: webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java (hunk 15 above)
Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: What happens is:
either,
- I send a request to join his network, and he sends back an AcceptJoin message
- Or I send an invite for him to join my network, and he sends back and AcceptJoin message

by "AcceptJoin" I intend both the "I just got invited to another network and I'm accepting" and the "Someone asked to enter my network and I'm accepting"

Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: That's because when you look at what happens in both cases, you see that they do the same thing:
both add the other person to their own routing table and then update it

Comment (francescobau) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: Oh ok now I think I get it... Thank you! :)

Comment (Cogno-Marco) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: So? Do you think I should write it better now that you understand?

Comment (francescobau) [reply webdictionary/src/main/java/com/eis0/kademlianetwork/ConnectionHandler.java:69]: Personally, I think you could actually write it better...
There can be people which are more retarded than me in this world...

**Comment by Cogno-Marco** [comment]
Ok, I think i've improved it now

**Review by francescobau — APPROVED** [review]
Much better! 👍
